### ISSUE TITLE
Drop multi-stage support

### ISSUE THREAD

**Issue by ninjapapa** [TresAmigosSD/SMV#1364]
The multiple stage function, especially the `SmvModuleLink` mechanism, was never adopted by clients. The recent project with multiple stages are simply independent projects under the same repo. 

The reason why this concept never been adopted could be the following:
- Too much for user to learn before using it
- User may want more explicit control on the output data of each "stage", instead of let SMV to manage for them
- Anyhow the multiple stages typically run in multiple steps, multi SmvApps might be more natural than a single SmvApp with multiple stages

We will remove `SmvModuleLink`, publish stage versions, etc. So effectively only support Single-stage project. 

In longer run, we may consider to allow multiple SmvApps under the same project.

**Comment by yw-yang** [comment]
I agree to remove `SmvModuleLink` to simplify the usage. The only concern is `SmvModuleLink` has been widely used in Telstra project metric implementation, is it okay to just remove it or we should inform Telstra side?

**Comment by ninjapapa** [comment]
@yw-yang good to know. 
Since switch between multiple published versions of the up-stream stage when run the down-stream stage is more a function development than production, is Client-T using that function after we delivered? Can we keep `SmvModuleLink` just as a pass through, but remove the config to multiple published version function? The code should still be able to run.

**Comment by yw-yang** [comment]
Exactly, `SmvModuleLink` used in Client-T is just a pass through. I think it's okay to remove the config to multiple published version function and make sure code is still able to run.

**Comment by yw-yang** [comment]
@ninjapapa I just opened a PR to keep `SmvModuleLink` as a pass through. I didn't remove the config to multiple published version function since I found a single module is also supported to be published out, shall we disable this behavior going forward?

**Comment by ninjapapa** [comment]
@yw-yang please see PR comments.

We should still keep publish itself, while removing the read back from published versions, since we do have publish-hive kept. We may remove those command line parameter after #1202

**Comment by yw-yang** [comment]
# Design

The aim of this ticket is to simplify the stage usage. To be backwards compatible, we will only remove the functionality that down-stream module can read back from up-stream stage published versions. This change will not impact any production code that used multi-stage and SmvModuleLink before.

So from user usage perspective, smv will
- still allow multiple stages in conf as `smv.stages=stage1, stage2`
- still allow `-s stage1` command line parameter to specify a stage level run
- still support `SmvModuleLink`, but not support read from published data function of it, so it is just a pass through
- Disable `smv.stages.stage1.version` usage in conf

I guess in the long run, we still want user to get rid of multiple stages instead of using multiple stages easier, so I don't think below should be covered, but let me know if any considerations.

- Allow direct cross stage module dependency, so stage2.M2 -> stage1.M1 without SmvModuleLink should be allowed, also stage1, M1 in this case does not need to be SmvOutput
- Extend the -s command line parameter to allow user to specify all levels in FQN hierarchy, for example -s stage1.etl, which will resolve all SmvOutput modules has FQN prefixed with stage1.etl

Instead I suggest to add a notification in [smv_stages.md](https://github.com/TresAmigosSD/SMV/blob/master/docs/user/smv_stages.md), e.g. this is a legacy functionality which is going to be removed in next major release.

**Comment by ninjapapa** [comment]
@yw-yang before we conclude that we will remove multiple stage totally, let's consider the result of this issue can achieve the best up to what we know. So in that sense, we need to make the `SmvModuleLink` concept as trivial as possible. The cross stage direct dependency is not some new function need to be introduced, instead it is to remove `SmvModuleLink`'s unnecessary logic.

Trivialize `SmvModuleLink` will allow us to simplify the entire module dependency logic. Even if it is not covered by this issue, it need to be done.

**Comment by yw-yang** [comment]
Sure, I am working on it and will cover these in the issue.

**Comment by yw-yang** [comment]
# Design updated

This issue will make below changes 
- disable smv.stages.stage1.version usage in conf
- disable versioned stage data read in
- allow direct cross stage module dependency and input module does not need to be SmvOutput
and will still keep
- allow multiple stages in conf as smv.stages=stage1, stage2
- allow -s stage1 command line parameter to specify a stage level run
- allow SmvModuleLink

For extending the -s command line parameter, opened another ticket #1372 to discuss the design.
